File: TypeScript/src/bot.ts
import { Board, Board2D } from "./board";

export class Bot extends Board {
  constructor(board: Board2D = [["#", "#", "#"], ["#", "#", "#"], ["#", "#", "#"]]) {
      super();
      this._board = board;
  }

  private _getEmptyTiles(board: Board2D): number[][] {
    let emptyTiles: number[][] = [];
    board
        .forEach((slice, y) => slice
        .forEach((tile, x) => tile === "#" 
            ? emptyTiles.push([x, y]) 
            : null
    ));
    return emptyTiles;
  };

  private _findBestMove(board: Board2D): [number, number] {
      let ret: [number /* Best value */, [number, number]] = [-100, [0, 0]];
      this._getEmptyTiles(board).forEach((tile) => {
          const [x, y] = tile;
          board[y][x] = "O";
          const value = this._minimax(board, false, 0);
          board[y][x] = "#";
          if(ret[0] < value) ret = [value, [x, y]];
      });
      return ret[1];
  };
    
  private _minimax(board: Board2D, isMaximizing = true, depth = 0): number {
    const state = this._determineBoardStatus(board);
    if(state === "loss") return 100 - depth;
    else if(state === "win") return -100 + depth;
    else if(state === "draw") return 0;
      
    let bestVal = isMaximizing ? -100 : 100;
    let player = isMaximizing ? "O" : "X";
    this._getEmptyTiles(board).forEach((tile) => {
        const [x, y] = tile;
        board[y][x] = player;
        let value = this._minimax([...board], !isMaximizing, depth + 1);
        bestVal = isMaximizing 
            ? Math.max(value, bestVal) 
            : Math.min(value, bestVal);
        board[y][x] = "#";
    });

    return bestVal;   
  };
    
  public getNextPlacement(): [number, number] {
    return this._findBestMove(this._board);
  }

}

File: Rust/src/bot.rs
use crate::board;
use std::cmp;

pub struct Bot {
    board: board::Board2d,
}

impl Bot {
    pub fn new() -> Bot {
        Bot {
            board: [['#'; 3]; 3],
        }
    }

    pub fn from(template_board: board::Board2d) -> Bot {
        Bot {
            board: template_board,
        }
    }
}

impl Bot {
    fn retrieve_empty_tiles(&self, &board: &board::Board2d) -> Vec<board::Coordinates> {
        let mut empty_tiles: Vec<board::Coordinates> = Vec::new();
        for y in 0..3 {
            for (x, tile) in board[y].iter().enumerate() {
                if *tile == '#' { empty_tiles.push((x as usize, y as usize)) };
            }
        }
        empty_tiles 
    }

    fn minimax(&self, board: &mut board::Board2d, is_maximizing: bool, depth: i32) -> i32 {
        let state: Option<i32> = match board::Board::from(*board).status() {
            board::BoardStatus::Win => Some(-100 + depth), 
            board::BoardStatus::Lose => Some(100 - depth),
            board::BoardStatus::Draw => Some(0), 
            board::BoardStatus::Undeclared => None
        };

        if let Some(s) = &state {
            return *s;
        };


        let mut best_val = if is_maximizing { -100 } else { 100 };
        let player = if is_maximizing { board::GamePiece::O } else { board::GamePiece::X };
        
        for (x, y) in self.retrieve_empty_tiles(&board) {
            board[y][x] = player.get_char();
            let value = self.minimax(board, !is_maximizing, depth+1);
            best_val = if is_maximizing { cmp::max(value, best_val) } else { cmp::min(value, best_val) };
            board[y][x] = '#';
        }

        best_val
    }

    pub fn find_best_move(&mut self) -> board::Coordinates {
        let mut coords: board::Coordinates = (0, 0);
        let mut best_val = -100;

        for (x, y) in self.retrieve_empty_tiles(&self.board) {
            self.board[y][x] = 'O';
            let value = self.minimax(&mut Box::new(self.board), false, 0);
            self.board[y][x] = '#';
            if best_val < value {
                coords = (x, y);
                best_val = value;
            }
        }

        coords
    }

}
